### DISCUSSION TITLE
Cache merging on ACF 1-to-1 relationships

### DISCUSSION THREAD

**Discussion by ptpittman** [wp-graphql/wp-graphql#3032] [Apollo GraphQL]
I'm using WPGraphQL with the latest ACF extension and Apollo GraphQL nextjs server side (app dir). All is working exceedingly well, but I'm getting one nagging error that I can't figure how best to shake. This doesn't have production impact particularly, it's just a warning, but I'd like to solve it as it shows up in every build.

Basically, there's a custom post type of "Author" that relates to a post as postAuthors, which has another field group inside of it of postAuthorFields.

so, in abstract simplified form,
```
 query Post {
    post {
      postAuthors {
        nodes {
          ... on PostAuthor {
            id
            title
            slug
            postAuthorFields {
              position
            }
          }
        }
      }
    }
  }
```

Obviously, as with all other types, there's different shapes to the data I pull out of this in different parts of the site. But for some reason, this is the only model with which Apollo complains about not having a unique identifier in postAuthorFields with which to merge in its cache.

Its output, basically, is:
```
Cache data may be lost when replacing the postAuthorFields field of a PostAuthor object.

This could cause additional (usually avoidable) network requests to fetch data that were otherwise cached.

To address this problem (which is not a bug in Apollo Client), either ensure all objects of type PostAuthorFields have an ID or a custom merge function, or define a custom merge function for the PostAuthor.postAuthorFields field, so InMemoryCache can safely merge these objects:
```

[(Full link to actual message here)](https://www.apollographql.com/docs/react/errors/#%7B%22version%22%3A%223.9.0%22%2C%22message%22%3A14%2C%22args%22%3A%5B%22postAuthorFields%22%2C%22PostAuthor%22%2C%22either%20ensure%20all%20objects%20of%20type%20PostAuthorFields%20have%20an%20ID%20or%20a%20custom%20merge%20function%2C%20or%20%22%2C%22PostAuthor.postAuthorFields%22%2C%22%7B%5Cn%20%20%5C%22__typename%5C%22%3A%20%5C%22PostAuthorFields%5C%22%2C%5Cn%20%20%5C%22clippedImage%5C%22%3A%20%7B%5Cn%20%20%20%20%5C%22__typename%5C%22%3A%20%5C%22AcfMediaItemConnectionEdge%5C%22%2C%5Cn%20%20%20%20%5C%22node%5C%22%3A%20%7B%5Cn%20%20%20%20%20%20%5C%22__typename%5C%22%3A%20%5C%22MediaItem%5C%22%2C%5Cn%20%20%20%20%20%20%5C%22mediaItemUrl%5C%22%3A%20%5C%22https%3A%2F%2Fmedia.pictongazette.ca%2Fwp-content%2Fuploads%2F2023%2F11%2F19165643%2FJason-Parks_cut.png%5C%22%2C%5Cn%20%20%20%20%20%20%5C%22mediaType%5C%22%3A%20%5C%22image%5C%22%2C%5Cn%20%20%20%20%20%20%5C%22mimeType%5C%22%3A%20%5C%22image%2Fpng%5C%22%2C%5Cn%20%20%20%20%20%20%5C%22sourceUrl%5C%22%3A%20%5C%22https%3A%2F%2Fmedia.pictongazette.ca%2Fwp-content%2Fuploads%2F2023%2F11%2F19165643%2FJason-Parks_cut.png%5C%22%2C%5Cn%20%20%20%20%20%20%5C%22slug%5C%22%3A%20%5C%22jason-parks_cut%5C%22%2C%5Cn%20%20%20%20%20%20%5C%22sizes(%7B%5C%5C%5C%22size%5C%5C%5C%22%3A%5C%5C%5C%22MEDIUM%5C%5C%5C%22%7D)%5C%22%3A%20%5C%22(max-width%3A%20255px)%20100vw%2C%20255px%5C%22%2C%5Cn%20%20%20%20%20%20%5C%22mediaDetails%5C%22%3A%20%7B%5Cn%20%20%20%20%20%20%20%20%5C%22__typename%5C%22%3A%20%5C%22MediaDetails%5C%22%2C%5Cn%20%20%20%20%20%20%20%20%5C%22sizes%5C%22%3A%20%5B%5Cn%20%20%20%20%20%20%20%20%20%20%7B%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%5C%22__typename%5C%22%3A%20%5C%22MediaSize%5C%22%2C%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%5C%22width%5C%22%3A%20%5C%22255%5C%22%2C%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%5C%22name%5C%22%3A%20%5C%22medium%5C%22%2C%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%5C%22sourceUrl%5C%22%3A%20%5C%22https%3A%2F%2Fmedia.pictongazette.ca%2Fwp-content%2Fuploads%2F2023%2F11%2F19165643%2FJason-Parks_cut-255x300.png%5C%22%2C%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%5C%22fileSize%5C%22%3A%2054937%2C%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%5C%22file%5C%22%3A%20%5C%22Jason-Parks_cut-255x300.png%5C%22%2C%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%5C%22height%5C%22%3A%20%5C%22300%5C%22%2C%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%5C%22mimeType%5C%22%3A%20%5C%22image%2Fpng%5C%22%5Cn%20%22%2C%22%7B%5Cn%20%20%5C%22__typename%5C%22%3A%20%5C%22PostAuthorFields%5C%22%2C%5Cn%20%20%5C%22authorPosition%5C%22%3A%20%5C%22Editor%5C%22%5Cn%7D%22%5D%7D)

It's possible I could write an elaborate custom merge specifically for this, but really it shouldn't be necessary - every postAuthorFields is 1-to-1 with a postAuthor, and that's its identifier (and there's no actual id for it selectable within the graph, nor should there really need to be). I'm curious if there's something about the ACF setup that might be triggering this only with these particular fields? Other 1-to-1s with field groups on other models, with similar variance in query shape, don't seem to trigger it.

**Comment by justlevine** [comment]
Total guess, but does it go away if you rename the CPT/ACF group (or GraphQL name)?

There's been several reports on Slack about WPGraphQL for ACF getting stumped when someone uses a reserved type (e.g. calling a Custom Taxonomy `Category` when that's used by built in post Categories). Could be that Author / postAuthor are just similar enough to be causing a small conflict internally without fully breaking the schema.

If not, I suggest you post in the #acf Slack channel or in the WPGraphQL for ACF repo for better visibility.
